galley/pkg/source/kube/source.go

```diff
@@ -0,0 +1,149 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kube
+
+import (
+	"sync"
+	"time"
+
+	"istio.io/istio/galley/pkg/runtime"
+	"istio.io/istio/galley/pkg/runtime/resource"
+	"istio.io/istio/galley/pkg/source/kube/builtin"
+	"istio.io/istio/galley/pkg/source/kube/client"
+	"istio.io/istio/galley/pkg/source/kube/dynamic"
+	kubeConverter "istio.io/istio/galley/pkg/source/kube/dynamic/converter"
+	"istio.io/istio/galley/pkg/source/kube/log"
+	"istio.io/istio/galley/pkg/source/kube/schema"
+
+	kubeDynamic "k8s.io/client-go/dynamic"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	// TODO(nmittler): Pilot allows specifying the namespace for the shared informer.
+	namespace = ""
+)
+
+// New creates a new kube source for the given schema.
+func New(interfaces client.Interfaces, resyncPeriod time.Duration, schema *schema.Instance,
+	cfg *kubeConverter.Config) (runtime.Source, error) {
+
+	var err error
+	var cl kubernetes.Interface
+	var dynClient kubeDynamic.Interface
+	var sharedInformers informers.SharedInformerFactory
+
+	sources := make([]runtime.Source, 0)
+	for i, spec := range schema.All() {
+		log.Scope.Infof("[%d]", i)
+		log.Scope.Infof("  Source:      %s", spec.CanonicalResourceName())
+		log.Scope.Infof("  Collection:  %s", spec.Target.Collection)
+
+		// If it's a known type, use a custom (optimized) source.
+		if builtin.IsBuiltIn(spec.Kind) {
+			// Lazy create the kube client.
+			if cl, err = getKubeClient(cl, interfaces); err != nil {
+				return nil, err
+			}
+			sharedInformers = getSharedInformers(sharedInformers, cl, resyncPeriod, namespace)
+
+			source, err := builtin.NewSource(sharedInformers, spec)
+			if err != nil {
+				return nil, err
+			}
+			sources = append(sources, source)
+		} else {
+			// Lazy-create the dynamic client
+			if dynClient, err = getDynamicClient(dynClient, interfaces); err != nil {
+				return nil, err
+			}
+			// Unknown types use the dynamic source.
+			source, err := dynamic.NewSource(dynClient, resyncPeriod, spec, cfg)
+			if err != nil {
+				return nil, err
+			}
+			sources = append(sources, source)
+		}
+	}
+
+	return &aggregate{
+		sources: sources,
+	}, nil
+}
+
+func getKubeClient(cl kubernetes.Interface, interfaces client.Interfaces) (kubernetes.Interface, error) {
+	if cl != nil {
+		return cl, nil
+	}
+	return interfaces.KubeClient()
+}
+
+func getSharedInformers(sharedInformers informers.SharedInformerFactory,
+	cl kubernetes.Interface, resyncPeriod time.Duration, namespace string) informers.SharedInformerFactory {
+	if sharedInformers != nil {
+		return sharedInformers
+	}
+	return informers.NewSharedInformerFactoryWithOptions(cl, resyncPeriod, informers.WithNamespace(namespace))
```

Comment:
It looks like we can simply remove the WithNamespace option and don't plumb in the constant, empty namespace.

Reply:
As discussed, I'd really prefer to make sure we have a good understanding of the code going in here. If we need it, we can certainly add it in the future.

Reply:
removed for now. We can re-introduce later if we think we need it.